### PR TITLE
backend/cuda: refactoring complex pack/unpack protocol switch in progress enqueue and minor bug fix

### PR DESCRIPTION
**Pull request by minsii**
## Pull Request Description
The current pack/unpack protocol switch code in `yaksuri_progress_enqueue` is complex and not easy to read. This patch refactors the long if-else based switch logic. The basic approach is to separate each situation (e.g., d2d_pack, d2rh_pack...) into subroutines so that later developers can easily find the place to optimize.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

**Comment by minsii** [comment]
test:yaksa/gpu

**Comment by minsii** [comment]
@hzhou would you be able to review this PR? It is mostly refactoring for better readability. I am doing more fixes/optimizations  based on the refactored version which I will post as separate PRs.

**Comment by minsii** [comment]
Just realized that the commit messages should be changed from `backend/cuda` to `backend`. No code change. Do we need test `ze`?

**Comment by hzhou** [comment]
> 
> 
> Just realized that the commit messages should be changed from `backend/cuda` to `backend`. No code change. Do we need test `ze`?

Ideally, yes. Unfortunately, that is difficult and adds a significant burden, so unless there are explicit changes to ZE-speicific code, I think we can relax the testing requirement.

File: src/backend/src/yaksuri_progress.c
Comment (hzhou) [inline src/backend/src/yaksuri_progress.c]: `d` means "device", right? What does `2` refer to?

File: src/backend/src/yaksuri_progress.c
Comment (hzhou) [inline src/backend/src/yaksuri_progress.c]: The name need to reflect this is for same-device transfer (is it?).

**Review by hzhou — COMMENTED** [review]
Generally looks good. Just a few minor comments

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c:20]: I will check the optimized build to make sure that these functions are inlined.

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: Yes, `d` means `device`. `2` refers to `to`. It tries to follow the naming of existing functions such as `pack_d2d_nop2p_acquire`

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: Not really. The current code only uses it for same-device fast path. But we will also use it to optimize the IPC and P2P-enabled cases in the upcoming optimization PR.

File: src/backend/src/yaksuri_progress.c
Comment (hzhou) [reply src/backend/src/yaksuri_progress.c]: I see. How about `set_subreq_pack_from_device`?

File: src/backend/src/yaksuri_progress.c
Comment (hzhou) [reply src/backend/src/yaksuri_progress.c]: okay

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c:20]: Confirmed that without `inline` the compiler can good job to inline those subroutines (tried gcc v7.5.0, assuming clang and icc can do better). Will remove the inline commit in the next push.

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: The proposed format may make some subroutine names very long, e.g.,`set_subreq_pack_from_registered_host`, `set_subreq_pack_from_device_to_device`, `set_subreq_pack_from_device_to_registered_host`. I think it is fine to use these abbreviation consistently in this file.

If you are OK with the abbreviation, I am not sure which is better between `set_pack_d2_subreq ` and `set_subreq_pack_d2`. What do you think?

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: An update for the naming of `singlechunk_pack`: after second read on this code, this routine is used only for `YAKSA_OP__REPLACE` in the current code. It is still unclear to me whether the same fast path can be used for other ops (e.g., MAX, MIN). But I guess it is OK to keep the name for now. Once we make optimization for non-REPLACE ops, we can adjust at that time.

**Comment by minsii** [comment]
This push deletes the inline commit

File: src/backend/src/yaksuri_progress.c
Comment (hzhou) [reply src/backend/src/yaksuri_progress.c]: We are more used to `d2d`, `d2h`, etc. although a short comment above the definition to expand the abbreviation can be helpful.

`_d2` is just too cryptic, mostly because we don't use such often. When used, it often means different from the meaning here. One of the heuristics (from "Clean code") is the name needs to be pronounceable and can be put into a sentence. `d2` doesn't make sense in a sentence.

Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: Agreed with your comment. But just the suggested name seems too long and a bit inconsistent with the other subroutines which use `d2d, d2h`. I guess the confusing part is we don't have the "destination" in `d2`. Will `d2any` read better?

Comment (hzhou) [reply src/backend/src/yaksuri_progress.c]: `from_dev` is 8 characters. `d2any` is 5 characters. Are we bargaining here? :)

Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: How about `from_registered_host` and `from_managed`? I am more concerning that it is not consistent with the other subroutines' name.

Comment (hzhou) [reply src/backend/src/yaksuri_progress.c]: Now you mentioned, were you using `m` to encode `managed`? It was very cryptic to me. The `from_xxx` routines are at one higher level than `_x2y` routines, I think the different name patterns are ok here -- the hierarchy is more important to read than the individuals. As for the two names, how about `from_reghost` (I'll accept `from_rhost` with some reluctance)? `from_managed` seems to be a minimal readable name to me.

**Review by hzhou — APPROVED** [review]
My comment on the naming (of 3 functions) remains, but I am not blocking the PR. I leave the decision back to you.

File: src/backend/src/yaksuri_progress.c
Comment (minsii) [reply src/backend/src/yaksuri_progress.c]: `from_rhost` seems good to me.

**Comment by minsii** [comment]
> My comment on the naming (of 3 functions) remains, but I am not blocking the PR. I leave the decision back to you.

Sorry I was tied with other work thus cannot follow up since yesterday night. Pushed changes for renaming. Will merge in after all jobs are cleaned.

**Comment by minsii** [comment]
test:yaksa/gpu

**Pull request merged**
